### PR TITLE
This code is all about pin features functionality

### PR DESCRIPTION
**Pull request by reyven-ai**

### CODE REVIEW

File: frontend/src/components/DropdownUtil/ProfileDropdown.tsx [hunk 1]
```diff
@@ -1,20 +1,22 @@
...
+  userId: number;
+}
+
+const ProfileUtil: React.FC<OtherComponentProps> = ({ username, userId }) => {
   const [isModalOpen, setIsModalOpen] = useState(false);
   const navigate = useNavigate();
 
+  // if (!userProfile) {
+  //   return <div>Profile not found</div>;
+  // }
```
Comment (artjom-mashanin) [inline frontend/src/components/DropdownUtil/ProfileDropdown.tsx:18]: Do you need this comment?

File: frontend/src/components/Layout/MainHeader.tsx [hunk 2]
```diff
@@ -56,41 +63,62 @@ function MainHeader() {
...
-            </li>
-            <li>
-              <ProfileUtil />
-            </li>
-          </ul>
+            </li> */}
+            {/* <li> */}
+            {/* <Chat /> */}
+            <SearchBar />
+            {/* </li> */}
```
Comment (artjom-mashanin) [inline frontend/src/components/Layout/MainHeader.tsx:114]: Do you need these comments?

File: frontend/src/components/pin/Pin.Details.tsx [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+                </li>
+                <li>
+                  <MenuDropdown />
+                </li>
+              </ul>
+              <SaveButton pin={pinDetails} />
+              {/* <button className="bg-backgroundButtonColor px-8 py-2 text-white rounded-[20px]">
+              {/* <button className="bg-backgroundButtonColor px-8 py-2 text-white rounded-[20px]">
+                Save
+              </button> */}
```
Comment (artjom-mashanin) [inline frontend/src/components/pin/Pin.Details.tsx:36]: Do you need these comments?

File: frontend/src/components/pin/Pin.dropzone.tsx [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+//         <div className="bg-[#e2e2e2] w-[370px] h-[450px] rounded-[30px] text-center border-2 border-dashed border-[#cbc9c9] flex items-center cursor-pointer">
+//           <p className="mx-[auto] w-[60%] flex flex-col items-center gap-[8px]">
+//             <ArrowUpwardRounded />
+//             Choose a file or drag and drop it here
+//           </p>
+//         </div>
+//       )}
+//     </div>
+//   );
+// };
```
Comment (artjom-mashanin) [inline frontend/src/components/pin/Pin.dropzone.tsx:94]: Do you need these comments?

File: frontend/src/hooks/useProfileAction.ts [hunk 5]
```diff
@@ -142,12 +149,35 @@ function transformProfileDataToInput(
...
+//     };
+
+//     fetchData();
+//     return () => {};
+//   }, [profileId]);
+
+//   return {
+//     profileDetails,
+//   };
+// };
```
Comment (artjom-mashanin) [inline frontend/src/hooks/useProfileAction.ts:173]: Do you need these comments?

File: frontend/src/pages/Profile/Profile.Details.tsx [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+//           </li>
+//         </ul>
+//         {activeTab === "created" && <PinProfile />}
+//         {activeTab === "saved" && <SaveList />}
+//       </div>
+//     </div>
+//   );
+// };
+
+// export default ProfileDetails;
```
Comment (artjom-mashanin) [inline frontend/src/pages/Profile/Profile.Details.tsx:78]: This whole file is commented out 
Do you need it?

**Review by artjom-mashanin — COMMENTED** [review]
Damn, that's a big PR :D 

Added some comments

File: dist/index.html
[the remark this answers is not in the document]
Comment (reyven-ai) [reply dist/index.html:1]: Done!

File: public/index.html
[the remark this answers is not in the document]
Comment (reyven-ai) [reply public/index.html:1]: Done!

File: frontend/src/components/DropdownUtil/ProfileDropdown.tsx (hunk 1 above)
Comment (reyven-ai) [reply frontend/src/components/DropdownUtil/ProfileDropdown.tsx:18]: Done!

File: frontend/src/components/Layout/MainHeader.tsx (hunk 2 above)
Comment (reyven-ai) [reply frontend/src/components/Layout/MainHeader.tsx:114]: Done!

File: frontend/src/components/pin/Pin.Details.tsx (hunk 3 above)
Comment (reyven-ai) [reply frontend/src/components/pin/Pin.Details.tsx:36]: Done!

File: frontend/src/hooks/useProfileAction.ts (hunk 5 above)
Comment (reyven-ai) [reply frontend/src/hooks/useProfileAction.ts:173]: Done!

File: frontend/src/components/pin/Pin.dropzone.tsx (hunk 4 above)
Comment (reyven-ai) [reply frontend/src/components/pin/Pin.dropzone.tsx:94]: Done!

File: frontend/src/pages/Profile/Profile.Details.tsx (hunk 6 above)
Comment (reyven-ai) [reply frontend/src/pages/Profile/Profile.Details.tsx:78]: Done!

**Review by artjom-mashanin — CHANGES_REQUESTED** [review]

File: frontend/package.json [hunk 7]
```diff
@@ -13,18 +13,24 @@
...
     "formik": "^2.4.5",
     "geolib": "^3.3.4",
     "react": "^18.2.0",
     "react-dom": "^18.2.0",
+    "react-dropzone": "^14.2.3",
     "react-query": "^3.39.3",
+    "react-redux": "^9.1.0",
     "react-router-dom": "^6.17.0",
+    "redux": "^5.0.1",
+    "redux-thunk": "^3.1.0",
```
[the remark this answers is not in the document]
Comment (reyven-ai) [reply frontend/package.json:30]: I just removed them

File: frontend/src/pages/Profile/Profile.Details.tsx (hunk 6 above)
Comment (reyven-ai) [reply frontend/src/pages/Profile/Profile.Details.tsx:78]: Done!

File: backend/src/database/migrations/20240208160712_save_pin.ts [hunk 8]
```diff
@@ -0,0 +1,18 @@
+import type { Knex } from "knex";
+
+export async function up(knex: Knex): Promise<void> {
+  return knex.schema.createTable("saved_pins", function (table) {
+    table.increments("id").primary();
+    table.integer("user_id").unsigned().references("user_id").inTable("users");
+    table.integer("pin_id").unsigned().references("id").inTable("pins");
+    table.string("title");
+    table.string("description");
+    table.string("image_url");
```
[the remark this answers is not in the document]
Comment (reyven-ai) [reply backend/src/database/migrations/20240208160712_save_pin.ts:10]: Done!

File: backend/src/modules/profile/profile.services.ts [hunk 9]
```diff
@@ -62,7 +62,7 @@ export async function getProfileDetails(
 ): Promise<ProfileData | null> {
   try {
     const userRepository = new UserDetailsRepository();
-    const getSelfUserProfile = await userRepository.getSelfUSerProfile(userId);
+    const getSelfUserProfile = await userRepository.getSelfUserProfile(userId);
```
[the remark this answers is not in the document]
Comment (reyven-ai) [reply backend/src/modules/profile/profile.services.ts:65]: Done!

File: backend/src/modules/profile/profile.services.ts [hunk 10]
```diff
@@ -62,7 +62,7 @@ export async function getProfileDetails(
 ): Promise<ProfileData | null> {
   try {
     const userRepository = new UserDetailsRepository();
```
[the remark this answers is not in the document]
Comment (reyven-ai) [reply backend/src/modules/profile/profile.services.ts:64]: Done!

File: backend/src/modules/users/users.route.ts [hunk 11]
```diff
@@ -24,4 +26,75 @@ router.get(
...
+      }
+
+      const data = {
+        userId,
+        pinId,
+        title: pin.title,
+        description: pin.description,
+        image_url: pin.image_url,
+        created_at: new Date(),
+        created_by: pin.created_by,
```
[the remark this answers is not in the document]
Comment (reyven-ai) [reply backend/src/modules/users/users.route.ts:51]: Done!

File: backend/src/modules/users/users.save.services.ts [hunk 12]
```diff
@@ -0,0 +1,48 @@
+import { PinData } from "../pin/pin.types";
+import { UserSavePinRepository } from "./users.save.repository";
+import { SavedData } from "./users.save.types";
+
+export async function savePin(data: SavedData): Promise<SavedData | null> {
```
[the remark this answers is not in the document]
Comment (reyven-ai) [reply backend/src/modules/users/users.save.services.ts:5]: Done!

File: frontend/src/validations/pin.validation.ts [hunk 13]
```diff
@@ -0,0 +1,14 @@
+import * as Yup from "yup";
+
+// The reason I dind't add any validation is because when ever the user is put some unacceptable file or size and trying to sudmit it they will get error from backend valdation and the error will be displayed on the screen
```
[the remark this answers is not in the document]
Comment (reyven-ai) [reply frontend/src/validations/pin.validation.ts]: I add some image validation for that

File: backend/src/modules/users/users.route.ts [hunk 14]
```diff
@@ -24,4 +26,75 @@ router.get(
   }
 );
 
+router.post(
+  "/:userId/savePin/",
+  checkAuthMiddleware,
+  async (req: Request, res: Response) => {
+    try {
+      const { pinId } = req.body;
+      const userId = req.params.userId;
```
[the remark this answers is not in the document]
Comment (reyven-ai) [reply backend/src/modules/users/users.route.ts]: Done!

File: backend/src/database/migrations/20240104221423_create_users_content.ts [hunk 15]
```diff
@@ -6,6 +6,9 @@ export async function up(knex: Knex): Promise<void> {
     table.integer("user_id").unsigned().references("user_id").inTable("users");
     table.text("description");
     table.string("image_url");
+    table.string("title");
+    table.string("link");
+    table.string("created_by").references("username").inTable("user_profiles");
```
[the remark this answers is not in the document]
Comment (artjom-mashanin) [reply backend/src/database/migrations/20240104221423_create_users_content.ts:11]: You are still changing existing migration — instead create a new migration that changes pins table and adds those new properties

File: backend/src/database/migrations/20240216202000_update_saved_pins.ts [hunk 16]
```diff
@@ -0,0 +1,19 @@
+import type { Knex } from "knex";
+
+export async function up(knex: Knex): Promise<void> {
+  return knex.schema.table("saved_pins", function (table) {
+    table.dropColumn("title");
+    table.dropColumn("description");
+    table.dropColumn("image_url");
+    table.dropColumn("created_by");
+  });
```
Comment (artjom-mashanin) [inline backend/src/database/migrations/20240216202000_update_saved_pins.ts:9]: Your previous migrattion for rewating the saved pins is not adding these properties anymore — so you don't need to drop the columns in here, cause they technically never existed

You would only ned tyo do that if you have merged the migration to main where you were creating those properties

So technically, you can just remove this migration

File: backend/src/modules/users/users.save.services.ts [hunk 17]
```diff
@@ -0,0 +1,46 @@
+import { PinData } from "../pin/pin.types";
+import { UserSavePinRepository } from "./users.save.repository";
+import { SavePinInput, SavedPinData } from "./users.save.types";
+
+export async function savePin(
+  data: SavePinInput
+): Promise<SavedPinData | null> {
+  try {
+    const pinRepository = new UserSavePinRepository();
+    const savedPin = await pinRepository.saveUserPin(
```
Comment (artjom-mashanin) [inline backend/src/modules/users/users.save.services.ts:10]: I'm curious, what would happen here if you're trying to save a pin that doesn't exist?
Or save it for a user that doesn't exist?

File: backend/src/database/migrations/20240216202000_update_saved_pins.ts (hunk 16 above)
Comment (reyven-ai) [reply backend/src/database/migrations/20240216202000_update_saved_pins.ts:9]: Done!

**Pull request merged**